src/value.rs

```diff
@@ -235,28 +239,58 @@ impl fmt::Debug for Value {
 
 impl fmt::Display for Value {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            Value::Unit => f.write_str("ε"),
-            Value::SumL(ref sub) => {
-                f.write_str("0")?;
-                if **sub != Value::Unit {
-                    write!(f, "{}", sub)
-                } else {
-                    Ok(())
+        #[derive(PartialEq, Eq)]
+        enum Parent {
+            Sum,
+            ProdL,
+            ProdR,
+        }
+
+        let mut parent_stack = vec![];
+        for val in self.pre_order_iter::<NoSharing>() {
+            match val {
+                Value::Unit => {
+                    if parent_stack.last() != Some(&Parent::Sum) {
+                        f.write_str("ε")?;
+                    }
+                    while let Some(parent) = parent_stack.pop() {
+                        if parent == Parent::ProdL {
+                            f.write_str(",")?;
+                            parent_stack.push(Parent::ProdR);
+                            break;
+                        } else if parent == Parent::ProdR {
+                            f.write_str(")")?;
+                        }
+                    }
                 }
-            }
-            Value::SumR(ref sub) => {
-                f.write_str("1")?;
-                if **sub != Value::Unit {
-                    write!(f, "{}", sub)
-                } else {
-                    Ok(())
+                Value::SumL(..) => {
+                    f.write_str("0")?;
+                    parent_stack.push(Parent::Sum);
+                }
+                Value::SumR(..) => {
+                    f.write_str("1")?;
+                    parent_stack.push(Parent::Sum);
+                }
+                Value::Prod(..) => {
+                    f.write_str("(")?;
+                    parent_stack.push(Parent::ProdL);
                 }
```

Comment:
522bf157cf79124b60f7bc346eaa9ee89b520471: This implementation is awfully bit-centric. `Left(Sum(Unit, Unit))` is rendered as `0(ε,ε)` for example. This is an artificial example, but don't we have these kinds of non-bit values in front of case expressions sometimes? I would render what we can as bits (or even bytes) and render the rest as `L` (left) and `R` (right). Since `ε` is usually the empty string, maybe we should use a different symbol for unit, like `●`.

We could open an issue and tackle this later.

Reply:
This is exactly what the existing code does, and it matches what the Bit Machine does internally.

Happy to open an issue to change it but it's separate from this PR.